src/Map.ts

```diff
@@ -0,0 +1,825 @@
+): <A>(d: Map<K, A>) => Type<F, [K, A]>
+export function toUnfoldable<K, F>(O: Ord<K>, unfoldable: Unfoldable<F>): <A>(d: Map<K, A>) => HKT<F, [K, A]>
+export function toUnfoldable<K, F>(O: Ord<K>, unfoldable: Unfoldable<F>): <A>(d: Map<K, A>) => HKT<F, [K, A]> {
+  const toArrayO = toArray(O)
+  return d => {
+    const arr = toArrayO(d)
+    const len = arr.length
+    return unfoldable.unfoldr(0, b => (b < len ? some(tuple(arr[b], b + 1)) : none))
+  }
+}
+
+/**
+ * Insert or replace a key/value pair in a Map
+ *
+ * @since 1.14.0
+ */
+export const insert = <K>(S: Setoid<K>): (<A>(k: K, a: A, m: Map<K, A>) => Map<K, A>) => {
+  const lookupS = lookupWithKey(S)
+  return (k, a, m) => {
+    const found = lookupS(k, m)
+    if (found.isNone()) {
+      const r = new Map(m)
+      r.set(k, a)
+      return r
+    } else if (found.value[1] !== a) {
+      const r = new Map(m)
+      r.set(found.value[0], a)
+      return r
+    }
+    return m
+  }
+}
+
+/**
+ * Delete a key and value from a map
+ *
+ * @since 1.14.0
+ */
+export const remove = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Map<K, A>) => {
+  const lookupS = lookupWithKey(S)
+  return (k, m) => {
+    const found = lookupS(k, m)
+    if (found.isSome()) {
+      const r = new Map(m)
+      m.delete(k)
+      return r
+    }
+    return m
+  }
+}
+
+/**
+ * Delete a key and value from a map, returning the value as well as the subsequent map
+ *
+ * @since 1.14.0
+ */
+export const pop = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Option<[A, Map<K, A>]>) => {
+  return (k, m) => {
+    const a = lookup(S)(k, m)
+    return a.isNone() ? none : some(tuple(a.value, remove(S)(k, m)))
+  }
+}
+
+/**
+ * Lookup the value for a key in a `Map`.
+ * If the result is a `Some`, the existing key is also returned.
+ * @since 1.14.0
+ */
+export const lookupWithKey = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Option<[K, A]>) => {
+  return <A>(k: K, m: Map<K, A>) => {
+    const entries = m.entries()
+    let e: IteratorResult<[K, A]>
+    while (!(e = entries.next()).done) {
+      const [ka, a] = e.value
+      if (S.equals(ka, k)) {
+        return some(tuple(k, a))
+      }
+    }
+    return none
+  }
+}
+
+/**
+ * Lookup the value for a key in a `Map`.
+ * @since 1.14.0
+ */
+export const lookup = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Option<A>) => {
+  const lookupWithKeyS = lookupWithKey(S)
+  return (k, m) => lookupWithKeyS(k, m).map(([_, a]) => a)
+}
+
+/**
+ * Test whether or not one Map contains all of the keys and values contained in another Map
+ *
+ * @since 1.14.0
+ */
+export const isSubmap = <K, A>(SK: Setoid<K>, SA: Setoid<A>): ((d1: Map<K, A>, d2: Map<K, A>) => boolean) => {
+  const lookupWithKeyS = lookupWithKey(SK)
+  return (d1: Map<K, A>, d2: Map<K, A>): boolean => {
+    const entries = d1.entries()
+    let e: IteratorResult<[K, A]>
+    while (!(e = entries.next()).done) {
+      const [k, a] = e.value
+      const d1OptA = lookupWithKeyS(k, d1)
```

Comment:
I think that this lookup can be removed, you can already read `[k, a]`

Reply:
I did this so that that `SK` (the `Setoid<K>`) would test both sets of keys for equality using the Setoid. I think it needs to stay like this?

Reply:
I changed this. Marking it as resolved.